mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxViewModelProvider.kt

```diff
@@ -22,68 +26,88 @@ object MvRxViewModelProvider {
 
-    @Suppress("UNCHECKED_CAST")
-    internal fun <VM : BaseMvRxViewModel<S>, S : MvRxState> createViewModel(
-        viewModelClass: Class<out VM>,
-        stateClass: Class<out S>,
-        viewModelContext: ViewModelContext,
-        stateRestorer: (S) -> S = { it },
-        initialStateFactory: MvRxStateFactory<VM, S> = RealMvRxStateFactory()
-    ): VM {
-        val initialState = initialStateFactory.createInitialState(viewModelClass, stateClass, viewModelContext, stateRestorer)
-        val factoryViewModel = viewModelClass.factoryCompanion()?.let { factoryClass ->
-            try {
-                factoryClass.getMethod("create", ViewModelContext::class.java, MvRxState::class.java)
-                    .invoke(factoryClass.instance(), viewModelContext, initialState) as VM?
-            } catch (exception: NoSuchMethodException) {
-                // Check for JvmStatic method.
-                viewModelClass.getMethod("create", ViewModelContext::class.java, MvRxState::class.java)
-                    .invoke(null, viewModelContext, initialState) as VM?
-            }
+        if (!savedStateRegistry.isRestored) {
+            error(ACCESSED_BEFORE_ON_CREATE_ERR_MSG)
         }
-        val viewModel = factoryViewModel ?: createDefaultViewModel(viewModelClass, initialState)
-        return requireNotNull(viewModel) {
-            // If null, use Kotlin reflect for best error message. We will crash anyway, so performance
-            // doesn't matter.
-            if (viewModelClass.kotlin.primaryConstructor?.parameters?.size?.let { it > 1 } == true) {
-                "${viewModelClass.simpleName} takes dependencies other than initialState. " +
-                    "It must have companion object implementing ${MvRxViewModelFactory::class.java.simpleName} " +
-                    "with a create method returning a non-null ViewModel."
-            } else {
-                "${viewModelClass::class.java.simpleName} must have primary constructor with a " +
-                    "single non-optional parameter that takes initial state of ${stateClass.simpleName}."
+
+        val stateRestorer = savedStateRegistry
+            .consumeRestoredStateForKey(key)
+            ?.toStateRestorer<S>(viewModelContext)
+
+        val restoredContext = stateRestorer?.viewModelContext ?: viewModelContext
+
+        val viewModel = ViewModelProvider(
+            viewModelContext.owner,
+            MvRxFactory(
+                viewModelClass,
+                stateClass,
+                restoredContext,
+                key,
+                stateRestorer?.toRestoredState,
+                forExistingViewModel,
+                initialStateFactory
+            )
+        ).get(key, viewModelClass)
+
+        try {
+            // Save the view model's state to the bundle so that it can be used to recreate
+            // state across system initiated process death.
+            viewModelContext.savedStateRegistry.registerSavedStateProvider(key) {
```

Comment:
Mostly curious, could you first call consume, and then call register to avoid the the try catch? Fine with this approach, though.

Reply:
Calling consume just tries to retrieve the bundle from the saved state, we would need to call `unregisterSavedStateProvider` instead. Calling unregister each time before calling register could work but would require storing the args which initialized the activity viewModel somewhere. This is because within a flow, using `existingViewModel` would unregister the fragment that initially created the view model and reregister the current fragment instead. We would no longer have access to the args that created the view model when saving to the bundle which is required when recreating the state after process death.